beyond-all-reason/Beyond-All-Reason
# Lua Select API, Lua based KeySelect & Customisable Smart Select

### Work done

Created a Lua based replacement for the existing hotkey selection code and added functionality to SmartSelect to utilise the same API to modify which units are selected when using drag select.

There were two main motivations for this:

- the existing C++ API can be inconsistent and buggy in places. As far as I can tell the NameContains and Category rules are broken, and the select count/one/part conclusions includes units that are already selected, which can be counter intuitive.

- writing it in Lua allows the code to be updated, maintained and improved more easily. Far more people should be able to work on the Lua code, and it should be easier to apply the changes.

An additional driver was being able to use the select syntax for SmartSelect commands in uikeys.txt.

The following filter was not implemented, I have no idea what it is meant to do:  `RulesParamEquals_<string>_<integer>`

I was advised that no one uses it anyway. Also an extra rule `AntiAir` was added for selecting Anti-air units.

The relevant files are:
- `select_api.lua` the common code used in both SmartSelect and KeySelect to decode selection command strings and cache the resulting select function
- `unit_smart_select.lua` this widget was modified to use the select api for selectbox commands in uikeys.txt
- `unit_key_select.lua` intended as a replacement for the existing select commands. This widget is disabled by default as we want users to test it before it is fully rolled out as a replacement.
- `compare_to_spring.lua` a test script written to identify differences between the existing C++ select implementation and the new Lua select implementation. It's not a proper unit/integration test as the results require a lot of interpretation.

See discord thread for more details: https://discord.com/channels/549281623154229250/1266331814440730677

#### Setup
The changes to SmartSelect are merged in directly. It will require modification to the uikeys.txt file to use. Something like this:

```
bind          Any+shift  selectbox Not_Aircraft
bind           Any+ctrl  selectbox Buildoptions
bind            Any+alt  selectbox Aircraft
bind          Any+space  selectbox Builder_Aircraft
```

To test the KeySelect open widgets and enable the KeySelect widget.

#### Test steps
SmartSelect
- Copy the above code into uikeys.txt and run `/keyreload`
- Create a variety of air/land units
- The `shift`, `ctrl`, `alt` and `space` modifiers should change which units are selected
- Check that the existing SmartSelect drag selection key binds still work as expected

KeySelect
- Run `/wigetselector` and enable `KeySelect`
- Check that all of your keybinds still work

### Known issues

There seems to be an issue where the `NameContain` filter doesn't work as expected if you have `ClearSelection` in the conclusion. It will centre on the unit, but not actually select it. I suspect it also happens for the `Category` filter. No idea why. Seeing as the `NameContain` and `Category` filters don't seem to work correctly in the existing version anyway, I don't see it as a deal breaker.
